### PR TITLE
Reloading remote schema

### LINKED ISSUE #284: Reloading remote schema
I am using hosted schema's (localhost python server) but I am having issues with reloading the schema's.

Is it possible to tell me how I can do that?
And if it is not possible yet: would it be possible to add such functionality?

### PR DESCRIPTION
**Pull request by angelozerr**
Reloading remote schema

Fixes #284

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR provides 2 new vscode XML commands.

To play with this PR, write XML like this:

```
<?xml-model href="http://maven.apache.org/xsd/maven-4.0.0.xsd" ?>
<project xmlns="http://maven.apache.org/POM/4.0.0" >
	<BAD_ELEMENT />
</project>
```

and open the 2 commands to check cache is evicted (on .lemminx/cache directory) and validation is working again after applying  the 2 commands (I mean validation and XML completion/hover based on XSD)

Here a demo:

XMLEvictCacheDemo

In this demo you can see that error disappear when command is executed and you have project root element which is higlighted in blue (because it says that XSD is downloading, but as it's too fast you can see the hover information about that). Once teh command is executed you see an information message that validation was done with success and validation should higlight BAD_ELEMENT as error. XML completion and hover based on XSD should work too. You can check that the maven.xsd file has changed on your local hard drive.

File: package.json [hunk 1]
```diff
@@ -25,7 +25,9 @@
     "onLanguage:xml",
     "onLanguage:xsl",
     "onCommand:xml.open.docs",
-    "onCommand:xml.open.docs.home"
+    "onCommand:xml.open.docs.home",
+    "onCommand:xml.validation.current.file",
```
Comment (datho7561) [inline package.json:29]: No need to add these here, since it doesn't make sense to try to validate a file before opening it.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: docs/Commands.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+This command opens the `XML Documentation`
+
+## Validate current XML file
+
+This command re-triggers the [XML Validation](Validation.md#xml-validation) for the current file.
+
+When the [Server Cache Path](Preferences.md#server-cache-path) is activated, the command removes the referenced XSD, DTD grammar from the local cache.
+
+## Validate current XML file
```
Comment (datho7561) [inline docs/Commands.md:17]: Validate all open XML files

File: src/extension.ts [hunk 3]
```diff
@@ -262,31 +262,52 @@ export function activate(context: ExtensionContext) {
...
-      });    
+      });
+
+      // Register custom XML commands
+      context.subscriptions.push(commands.registerCommand(Commands.VALIDATE_CURRENT_FILE, async (params) => {
+        const uri = window.activeTextEditor.document.uri;
+        const identifier = TextDocumentIdentifier.create(uri.toString());
+        commands.executeCommand(Commands.EXECUTE_WORKSPACE_COMMAND, Commands.VALIDATE_CURRENT_FILE, identifier).
+          then(() => {
+            window.showInformationMessage('Validate current XML file done with success!');
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/extension.ts]: fixed

File: src/extension.ts [hunk 4]
```diff
@@ -262,31 +262,52 @@ export function activate(context: ExtensionContext) {
...
+          }, error => {
+            window.showErrorMessage('Error while XML validation: ' + error.message);
+          });
+      }));
+      context.subscriptions.push(commands.registerCommand(Commands.VALIDATE_ALL_FILES, async () => {
+        commands.executeCommand(Commands.EXECUTE_WORKSPACE_COMMAND, Commands.VALIDATE_ALL_FILES).
+          then(() => {
+            window.showInformationMessage('Validate all opened XML files done with success!');
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/extension.ts]: fixed

File: src/extension.ts [hunk 5]
```diff
@@ -262,31 +262,52 @@ export function activate(context: ExtensionContext) {
...
+          }, error => {
+            window.showErrorMessage('Error while XML validation: ' + error.message);
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/extension.ts]: fixed

File: package.json (hunk 1 above)
Comment (angelozerr) [reply package.json]: fixed

File: src/extension.ts [hunk 6]
```diff
@@ -262,31 +262,52 @@ export function activate(context: ExtensionContext) {
...
+
+      // Register custom XML commands
+      context.subscriptions.push(commands.registerCommand(Commands.VALIDATE_CURRENT_FILE, async (params) => {
+        const uri = window.activeTextEditor.document.uri;
+        const identifier = TextDocumentIdentifier.create(uri.toString());
+        commands.executeCommand(Commands.EXECUTE_WORKSPACE_COMMAND, Commands.VALIDATE_CURRENT_FILE, identifier).
+          then(() => {
+            window.showInformationMessage('Validate current XML file done with success!');
+          }, error => {
+            window.showErrorMessage('Error while XML validation: ' + error.message);
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/extension.ts]: fixed

File: docs/Commands.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+This command re-triggers the [XML Validation](Validation.md#xml-validation) for the all opened XML files.
+
+When the [Server Cache Path](Preferences.md#server-cache-path) is activated, the command removes all XSD, DTD grammars from the local cache.
```
Comment (datho7561) [inline docs/Commands.md:21]: Does it remove all the grammars, or just the grammars referenced by the opened files?

File: src/extension.ts [hunk 8]
```diff
@@ -262,31 +262,52 @@ export function activate(context: ExtensionContext) {
...
+          then(() => {
+            window.showInformationMessage('The current XML file was successfully validated.');
+          }, error => {
+            window.showErrorMessage('Error during XML validation ' + error.message);
+          });
+      }));
+      context.subscriptions.push(commands.registerCommand(Commands.VALIDATE_ALL_FILES, async () => {
+        commands.executeCommand(Commands.EXECUTE_WORKSPACE_COMMAND, Commands.VALIDATE_ALL_FILES).
+          then(() => {
+            window.showInformationMessage('All open XML files was successfully validated.');
```
Comment (datho7561) [inline src/extension.ts:281]: All open XML files were successfully validated.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: docs/Commands.md (hunk 7 above)
Comment (angelozerr) [reply docs/Commands.md:21]: all grammars

File: src/extension.ts (hunk 8 above)
Comment (angelozerr) [reply src/extension.ts]: fixed

File: docs/Commands.md (hunk 2 above)
Comment (angelozerr) [reply docs/Commands.md]: fixed

File: src/extension.ts [hunk 9]
```diff
@@ -262,31 +262,52 @@ export function activate(context: ExtensionContext) {
...
         return await commands.executeCommand(params.command, ...params.arguments);
-      });    
+      });
+
+      // Register custom XML commands
+      context.subscriptions.push(commands.registerCommand(Commands.VALIDATE_CURRENT_FILE, async (params) => {
+        const uri = window.activeTextEditor.document.uri;
+        const identifier = TextDocumentIdentifier.create(uri.toString());
+        commands.executeCommand(Commands.EXECUTE_WORKSPACE_COMMAND, Commands.VALIDATE_CURRENT_FILE, identifier).
+          then(() => {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/extension.ts:272]: I add a condition in package.json with menus to enable this command only when file has xml language id. I think it's enough, no?

[the remark this answers is not in the document]
Comment (datho7561) [reply src/extension.ts:272]: Yes, I think that's enough.

File: docs/Commands.md (hunk 7 above)
Comment (datho7561) [reply docs/Commands.md:21]: In this case, then I think the command description should same something like "clear remote grammar cache" to make it clear that it clears all the cached grammars.

Comment (datho7561) [reply docs/Commands.md:21]: Or "Clear remote grammar cache and revalidate all opened files"

Comment (angelozerr) [reply docs/Commands.md]: fixed

File: docs/Commands.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+## Open XML Documentation
+
+This command opens the `XML Documentation`
+
+## Validate current XML file
+
+This command re-triggers the [XML Validation](Validation.md#xml-validation) for the current file.
+
+When the [Server Cache Path](Preferences.md#server-cache-path) is activated, the command removes the referenced XSD, DTD grammar from the local cache.
```
Comment (angelozerr) [inline docs/Commands.md:15]: @datho7561 do you think the description is understandable?

Comment (datho7561) [reply docs/Commands.md:15]: This description makes sense. The description in the package.json is the one that I think needs improving.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: package.json [hunk 11]
```diff
@@ -379,8 +379,26 @@
...
         "category": "XML"
+      },
+      {
+        "command": "xml.validation.current.file",
+        "title": "Validate current XML file",
+        "category": "XML"
+      },
+      {
+        "command": "xml.validation.all.files",
+        "title": "Validate all opened XML files",
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply package.json:390]: I though that but `Validate all opened XML files and evicts all the cache` is a little long, no?

If we add that, we should update too the other command like `Validate current XML file and evicts referenced grammar cache` which is long too, no?

[the remark this answers is not in the document]
Comment (datho7561) [reply package.json:390]: I think its kind of long, but it does fit in the command bar:

Screenshot from 2020-10-15 15-53-47

IMHO, I think a more precise description is more valuable then keeping the name short.

[the remark this answers is not in the document]
Comment (datho7561) [reply package.json:390]: @angelozerr  What about "Revalidate current file" "Revalidate all opened files"?

[the remark this answers is not in the document]
Comment (angelozerr) [reply package.json]: fixed

**Review by datho7561 — APPROVED** [review]

**Pull request merged**